controllers/scopetemplate_controller.go

```diff
@@ -88,5 +111,122 @@ func (r *ScopeTemplateReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 func (r *ScopeTemplateReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&operatorsv1.ScopeTemplate{}).
+		// Set up a watch for ScopeInstance to handle requeuing of requests for ScopeTemplate
+		Watches(&source.Kind{Type: &operatorsv1.ScopeInstance{}}, handler.EnqueueRequestsFromMapFunc(r.mapToScopeTemplate)).
 		Complete(r)
 }
+
+func (r *ScopeTemplateReconciler) mapToScopeTemplate(obj client.Object) (requests []reconcile.Request) {
+	if obj == nil || obj.GetName() == "" {
+		return
+	}
+
+	ctx := context.TODO()
+	//(todo): Check if obj can be converted into a scope instance.
+	scopeInstance := &operatorsv1.ScopeInstance{}
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: obj.GetName()}, scopeInstance); err != nil {
+		return nil
+	}
+
+	// Exit early if scopeInstance doesn't reference a scopeTemplate
+	if scopeInstance.Spec.ScopeTemplateName == "" {
+		return nil
+	}
+
+	// enqueue requests for ScopeTemplate based on Name and Namespace
+	request := reconcile.Request{
+		NamespacedName: types.NamespacedName{Namespace: obj.GetNamespace(), Name: scopeInstance.Spec.ScopeTemplateName},
+	}
+
+	requests = append(requests, request)
+
+	return requests
+}
+
+func (r *ScopeTemplateReconciler) ensureClusterRoles(ctx context.Context, st *operatorsv1.ScopeTemplate) error {
+	scopeinstances := operatorsv1.ScopeInstanceList{}
+
+	if err := r.Client.List(ctx, &scopeinstances, client.InNamespace(st.Namespace)); err != nil {
+		return fmt.Errorf("list scope instances: %v", err)
+	}
+
+	// Create all the ClusterRoles
+	for i := range scopeinstances.Items {
+		sInstance := scopeinstances.Items[i]
+		if sInstance.Spec.ScopeTemplateName == st.Name {
+			log.Log.Info("ScopeInstance found that references ScopeTemplate", "name", st.Name)
+			for _, cr := range st.Spec.ClusterRoles {
+				clusterRole := &rbacv1.ClusterRole{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: cr.GenerateName,
+						OwnerReferences: []metav1.OwnerReference{{
+							APIVersion: st.APIVersion,
+							Kind:       st.Kind,
+							Name:       st.GetObjectMeta().GetName(),
+							UID:        st.GetObjectMeta().GetUID(),
+						}},
+						Labels: map[string]string{
+							scopeTemplateUIDKey:    string(st.GetUID()),
+							scopeTemplateHashKey:   util.HashObject(st.Spec),
+							clusterRoleGenerateKey: cr.GenerateName,
+						},
+					},
+					Rules: cr.Rules,
+				}
+
+				crbList := &rbacv1.ClusterRoleList{}
+				if err := r.Client.List(ctx, crbList, client.MatchingLabels{
+					scopeTemplateUIDKey:    string(st.GetUID()),
+					clusterRoleGenerateKey: cr.GenerateName,
+				}); err != nil {
+					return err
+				}
+
+				if len(crbList.Items) > 1 {
+					return fmt.Errorf("more than one ClusterRole found %s", cr.GenerateName)
```

Comment:
We probably want to do something smarter here, let's add a todo here to compare the clusterRoles against the expected values. Does not need to be a part of this PR.

Reply:
Added.